File: LC103.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        boolean reverse = false;
        Deque<TreeNode> deque = new ArrayDeque<>();
        List<List<Integer>> res = new ArrayList<>();
        if (root != null) deque.offerLast(root);
        while (!deque.isEmpty()) {
            int size = deque.size();
            List<Integer> list = new ArrayList<>();
            for (int i = 0; i < size; i++) {
                TreeNode node;
                if (reverse) node = deque.pollLast();
                else node = deque.pollFirst();
                if (node == null) continue;
                list.add(node.val);
                if (reverse) {
                    if (node.right != null) deque.offerFirst(node.right);
                    if (node.left != null) deque.offerFirst(node.left);
                }else{
                    if (node.left != null) deque.offerLast(node.left);
                    if (node.right != null) deque.offerLast(node.right);
                }

            }
            res.add(list);
            reverse = !reverse;
        }
        return res;
    }
}

File: src/LC103.go
package main

import "container/list"

//
//type TreeNode struct {
//	Val   int
//	Left  *TreeNode
//	Right *TreeNode
//}

func zigzagLevelOrder(root *TreeNode) [][]int {
	var res [][]int
	if root == nil {
		return res
	}
	var reverse bool
	reverse = false
	l := list.New()
	l.PushBack(root)

	for l.Len() != 0 {
		len := l.Len()
		temparr := list.New()
		for i := 0; i < len; i++ {
			temp := l.Front()
			temparr.PushBack(temp.Value.(*TreeNode).Val)
			l.Remove(temp)
			if temp.Value.(*TreeNode).Left != nil {
				l.PushBack(temp.Value.(*TreeNode).Left)
			}
			if temp.Value.(*TreeNode).Right != nil {
				l.PushBack(temp.Value.(*TreeNode).Right)
			}
		}
		var arr []int
		if reverse {
			for temparr.Len() > 0 {
				arr = append(arr, temparr.Back().Value.(int))
				temparr.Remove(temparr.Back())
			}

		} else {
			for temparr.Len() > 0 {
				arr = append(arr, temparr.Front().Value.(int))
				temparr.Remove(temparr.Front())
			}
		}
		reverse = !reverse
		res = append(res, arr)
	}
	return res
}
